syncthing-discovery/src/main/kotlin/net/syncthing/java/discovery/protocol/GlobalDiscoveryHandler.kt

```diff
@@ -14,125 +14,78 @@
-                    HttpStatus.SC_NOT_FOUND -> {
-                        logger.debug("device not found: {}", deviceId)
-                        return@execute emptyList()
-                    }
-                    HttpStatus.SC_OK -> {
-                        val announcementMessage = AnnouncementMessage.parse(
-                                JsonReader(
-                                        StringReader(
-                                                EntityUtils.toString(response.entity)
-                                        )
-                                )
-                        )
-                        return@execute (announcementMessage.addresses)
-                                .map { DeviceAddress(deviceId.deviceId, it) }
-                    }
-                    else -> {
-                        throw IOException("http error ${response.statusLine}, response ${EntityUtils.toString(response.entity)}")
+        return coroutineScope {
+            deviceIds
+                    .distinct()
+                    .map { deviceId ->
+                        async {
+                            queryAnnounceServers(
+                                    servers = discoveryServers,
+                                    deviceId = deviceId
+                            )
+                        }
                     }
-                }
-            }
-        } catch (e: Exception) {
-            when (e) {
-                is IOException, is NoSuchAlgorithmException, is KeyStoreException, is KeyManagementException ->
-                    throw IOException(e)
-                else -> throw e
-            }
+                    .map { it.await() }
+                    .flatten()
         }
     }
 
-    override fun close() {}
-
-    private data class AnnouncementMessage(val addresses: List<String>) {
-        companion object {
-            private const val ADDRESSES = "addresses"
+    suspend fun query(deviceId: DeviceId) = queryAnnounceServers(
+            servers = pickAnnounceServers(),
+            deviceId = deviceId
+    )
 
-            fun parse(reader: JsonReader): AnnouncementMessage {
-                var addresses = listOf<String>()
+    suspend fun pickAnnounceServers() = AddressRanker
+            .pingAddresses(configuration.discoveryServers.map { DeviceAddress(it, "tcp://$it:443") })
+            .map { it.deviceId }
 
-                reader.beginObject()
-                while (reader.hasNext()) {
-                    when (reader.nextName()) {
-                        ADDRESSES -> {
-                            val newAddresses = ArrayList<String>()
+    suspend fun queryAnnounceServers(servers: List<String>, deviceId: DeviceId) = coroutineScope {
+        servers
+                .map { server ->
+                    async {
+                        try {
+                            queryAnnounceServer(server, deviceId)
+                        } catch (ex: Exception) {
```

Comment:
Only catch specific exceptions.

Reply:
This would make the whole query fail (and ignore the result from other discovery servers). This would not be very helpful here.

Reply:
Got it. I rethrow now when it's not an IOException, DeviceNotFoundException or TooManyRequestsException